File: experiments-local/experiment-18/runs/agent=omp_language=typescript_model=openrouter/anthropic/claude-sonnet-4.6_tooling=none/rep1/src/loader.ts
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";
import { parse } from "csv-parse/sync";
import type { Match, Player } from "./types.js";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const DATA_DIR = path.resolve(__dirname, "../data/kaggle");

// Strip state suffix like "Palmeiras-SP" → "Palmeiras"
function normalizeTeamName(raw: string): string {
  if (!raw) return raw;
  return raw.replace(/\s*-\s*[A-Z]{2}$/, "").trim();
}

// Parse multiple date formats → ISO YYYY-MM-DD
function parseDate(raw: string): string {
  if (!raw) return "";
  const s = raw.trim();
  // Brazilian DD/MM/YYYY
  const br = s.match(/^(\d{1,2})\/(\d{1,2})\/(\d{4})/);
  if (br) return `${br[3]}-${br[2].padStart(2, "0")}-${br[1].padStart(2, "0")}`;
  // ISO with optional time
  const iso = s.match(/^(\d{4}-\d{2}-\d{2})/);
  if (iso) return iso[1];
  return s.substring(0, 10);
}

function toInt(v: unknown): number {
  const n = parseInt(String(v), 10);
  return isNaN(n) ? 0 : n;
}

function toFloat(v: unknown): number {
  const n = parseFloat(String(v));
  return isNaN(n) ? 0 : n;
}

function readCsv(filename: string): Record<string, string>[] {
  const file = path.join(DATA_DIR, filename);
  const content = fs.readFileSync(file);
  return parse(content, {
    columns: true,
    skip_empty_lines: true,
    relax_quotes: true,
    trim: true,
    bom: true,
  }) as Record<string, string>[];
}

function loadBrasileirao(): Match[] {
  const rows = readCsv("Brasileirao_Matches.csv");
  return rows.map((r) => ({
    date: parseDate(r.datetime),
    competition: "Brasileirao" as const,
    season: toInt(r.season),
    round: String(r.round),
    homeTeam: normalizeTeamName(r.home_team),
    awayTeam: normalizeTeamName(r.away_team),
    homeGoals: toInt(r.home_goal),
    awayGoals: toInt(r.away_goal),
    homeState: r.home_team_state,
    awayState: r.away_team_state,
  }));
}

function loadCopaDoBrasil(): Match[] {
  const rows = readCsv("Brazilian_Cup_Matches.csv");
  return rows.map((r) => ({
    date: parseDate(r.datetime),
    competition: "Copa do Brasil" as const,
    season: toInt(r.season),
    round: String(r.round),
    homeTeam: normalizeTeamName(r.home_team),
    awayTeam: normalizeTeamName(r.away_team),
    homeGoals: toInt(r.home_goal),
    awayGoals: toInt(r.away_goal),
  }));
}

function loadLibertadores(): Match[] {
  const rows = readCsv("Libertadores_Matches.csv");
  return rows.map((r) => ({
    date: parseDate(r.datetime),
    competition: "Libertadores" as const,
    season: toInt(r.season),
    stage: r.stage,
    homeTeam: normalizeTeamName(r.home_team),
    awayTeam: normalizeTeamName(r.away_team),
    homeGoals: toInt(r.home_goal),
    awayGoals: toInt(r.away_goal),
  }));
}

function loadExtended(): Match[] {
  const rows = readCsv("BR-Football-Dataset.csv");
  return rows.map((r) => ({
    date: parseDate(r.date),
    competition: "Extended" as const,
    season: r.date ? toInt(r.date.split("-")[0]) : 0,
    round: r.tournament,
    homeTeam: normalizeTeamName(r.home),
    awayTeam: normalizeTeamName(r.away),
    homeGoals: toFloat(r.home_goal),
    awayGoals: toFloat(r.away_goal),
    homeCorners: toFloat(r.home_corner),
    awayCorners: toFloat(r.away_corner),
    homeShots: toFloat(r.home_shots),
    awayShots: toFloat(r.away_shots),
  }));
}

function loadHistorical(): Match[] {
  const rows = readCsv("novo_campeonato_brasileiro.csv");
  return rows.map((r) => ({
    date: parseDate(r.Data),
    competition: "Brasileirao" as const,
    season: toInt(r.Ano),
    round: String(r.Rodada),
    homeTeam: normalizeTeamName(r.Equipe_mandante),
    awayTeam: normalizeTeamName(r.Equipe_visitante),
    homeGoals: toInt(r.Gols_mandante),
    awayGoals: toInt(r.Gols_visitante),
    homeState: r.Mandante_UF,
    awayState: r.Visitante_UF,
    arena: r.Arena,
  }));
}

function loadPlayers(): Player[] {
  const rows = readCsv("fifa_data.csv");
  return rows.map((r) => ({
    id: toInt(r.ID),
    name: r.Name?.trim() ?? "",
    age: toInt(r.Age),
    nationality: r.Nationality?.trim() ?? "",
    overall: toInt(r.Overall),
    potential: toInt(r.Potential),
    club: r.Club?.trim() ?? "",
    position: r.Position?.trim() ?? "",
    jerseyNumber: r["Jersey Number"] ? toInt(r["Jersey Number"]) : undefined,
    height: r.Height?.trim(),
    weight: r.Weight?.trim(),
    crossing: r.Crossing ? toInt(r.Crossing) : undefined,
    finishing: r.Finishing ? toInt(r.Finishing) : undefined,
    dribbling: r.Dribbling ? toInt(r.Dribbling) : undefined,
    shortPassing: r.ShortPassing ? toInt(r.ShortPassing) : undefined,
    longPassing: r.LongPassing ? toInt(r.LongPassing) : undefined,
    ballControl: r.BallControl ? toInt(r.BallControl) : undefined,
    acceleration: r.Acceleration ? toInt(r.Acceleration) : undefined,
    sprintSpeed: r.SprintSpeed ? toInt(r.SprintSpeed) : undefined,
    stamina: r.Stamina ? toInt(r.Stamina) : undefined,
    strength: r.Strength ? toInt(r.Strength) : undefined,
    reactions: r.Reactions ? toInt(r.Reactions) : undefined,
    composure: r.Composure ? toInt(r.Composure) : undefined,
    gkDiving: r.GKDiving ? toInt(r.GKDiving) : undefined,
    gkHandling: r.GKHandling ? toInt(r.GKHandling) : undefined,
    gkKicking: r.GKKicking ? toInt(r.GKKicking) : undefined,
    gkPositioning: r.GKPositioning ? toInt(r.GKPositioning) : undefined,
    gkReflexes: r.GKReflexes ? toInt(r.GKReflexes) : undefined,
  }));
}

// Singleton cache — loaded once on first access
let _matches: Match[] | null = null;
let _players: Player[] | null = null;

export function getAllMatches(): Match[] {
  if (_matches) return _matches;
  _matches = [
    ...loadBrasileirao(),
    ...loadCopaDoBrasil(),
    ...loadLibertadores(),
    ...loadExtended(),
    ...loadHistorical(),
  ];
  return _matches;
}

export function getAllPlayers(): Player[] {
  if (_players) return _players;
  _players = loadPlayers();
  return _players;
}

// For tests: inject mock data
export function _setMatchesForTest(matches: Match[]): void {
  _matches = matches;
}
export function _setPlayersForTest(players: Player[]): void {
  _players = players;
}
export function _resetCache(): void {
  _matches = null;
  _players = null;
}

File: experiments-local/experiment-mu-glm53-thinking/runs/agent=oc-think-off_language=python_model=openrouter/z-ai/glm-5.3_tooling=none/rep3/brazilian_soccer/loader.py
"""Data loading and normalization for the Brazilian soccer datasets.

Handles all 6 CSV files from data/kaggle/, normalizing team names,
dates, and scores into a common Match record so queries can be run
uniformly across competitions and files.

Team-name matching is two-tier:
- normalize_name() strips accents/parentheses and lowercases, but keeps
  state suffixes so 'Atletico-MG' and 'Atletico-PR' stay distinct.
- name_matches() lets a suffix-less query ('palmeiras') match any
  suffixed variant ('palmeiras-sp'), while a suffixed query
  ('atletico-mg') matches only that exact team.
"""

from __future__ import annotations

import csv
import re
import unicodedata
from collections import defaultdict
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Optional

DEFAULT_DATA_DIR = Path(__file__).resolve().parent.parent / "data" / "kaggle"

BRASILEIRAO = "Brasileirão"
COPA_DO_BRASIL = "Copa do Brasil"
LIBERTADORES = "Copa Libertadores"
BR_DATASET = "BR Football Dataset"
HISTORICO = "Brasileirão (2003-2019)"

# Normalized competition names that refer to the same real-world competition.
COMPETITION_ALIASES: dict[str, set[str]] = {
    "brasileirao": {
        "brasileirao",
        "serie a",
        "brasileirao (2003-2019)",
        "campeonato brasileiro serie a",
    },
    "copa do brasil": {"copa do brasil"},
    "libertadores": {"copa libertadores", "libertadores"},
    "serie b": {"serie b"},
}

# When several files cover the same competition+season, prefer the most
# authoritative/dedicated source.
_SOURCE_PRIORITY = [
    "Brasileirao_Matches.csv",
    "Brazilian_Cup_Matches.csv",
    "Libertadores_Matches.csv",
    "novo_campeonato_brasileiro.csv",
    "BR-Football-Dataset.csv",
]

_SUFFIX_RE = re.compile(r"\s*-\s*[a-z]{2,3}\s*$")


def normalize_name(name: str) -> str:
    """Normalize a team/player/competition name for comparisons.

    Strips accents, removes parenthetical annotations, lowercases, and
    collapses whitespace. State suffixes like '-SP' are kept so that
    distinct teams (Atletico-MG vs Atletico-PR) remain distinct; use
    base_name()/name_matches() for suffix-insensitive matching.
    """
    if not name:
        return ""
    s = unicodedata.normalize("NFKD", name)
    s = "".join(c for c in s if not unicodedata.combining(c))
    s = re.sub(r"\([^)]*\)", " ", s)
    s = re.sub(r"\s+", " ", s).strip().lower()
    return s


def base_name(normalized: str) -> str:
    """Strip a trailing state suffix from an already-normalized name."""
    return _SUFFIX_RE.sub("", normalized).strip()


def name_matches(query: str, target: str) -> bool:
    """True if a user-supplied query name matches a dataset team name.

    - A query with a suffix ('Atletico-MG') matches only that team.
    - A query without a suffix ('Palmeiras') matches any state variant
      ('Palmeiras-SP', 'Palmeiras').
    """
    qn = normalize_name(query)
    tn = normalize_name(target)
    if not qn or not tn:
        return False
    if qn == tn:
        return True
    # suffix-less query matches suffixed target
    if _SUFFIX_RE.sub("", qn) == qn and base_name(tn) == qn:
        return True
    return False


def display_name(name: str) -> str:
    """Return a clean display name (no state suffix, no parentheses)."""
    s = name.strip()
    s = re.sub(r"\([^)]*\)", "", s)
    s = re.sub(r"\s*-\s*[A-Z]{2,3}\s*$", "", s)
    return re.sub(r"\s+", " ", s).strip()


def competition_group(name: str) -> set[str]:
    """Resolve a competition name to its set of normalized aliases.

    Falls back to a literal normalized name if no alias group matches.
    """
    n = normalize_name(name)
    for aliases in COMPETITION_ALIASES.values():
        if n in aliases or base_name(n) in aliases:
            return set(aliases)
    return {n}


_DATE_FORMATS = (
    "%Y-%m-%d %H:%M:%S",
    "%Y-%m-%d",
    "%d/%m/%Y",
)


def parse_date(value: str) -> Optional[datetime]:
    """Parse dates in ISO or Brazilian (DD/MM/YYYY) formats."""
    if not value:
        return None
    value = value.strip()
    for fmt in _DATE_FORMATS:
        try:
            return datetime.strptime(value, fmt)
        except ValueError:
            continue
    return None


def _to_int(value) -> Optional[int]:
    if value is None:
        return None
    s = str(value).strip()
    if s == "" or s.lower() == "nan":
        return None
    try:
        return int(float(s))
    except ValueError:
        return None


@dataclass
class Match:
    """A single match, normalized across all source files."""

    competition: str
    season: Optional[int]
    date: Optional[datetime]
    home: str
    away: str
    home_goal: Optional[int]
    away_goal: Optional[int]
    round: Optional[str] = None
    stage: Optional[str] = None
    stadium: Optional[str] = None
    home_state: Optional[str] = None
    away_state: Optional[str] = None
    source_file: str = ""
    # extras from BR-Football-Dataset
    corners: Optional[dict] = field(default=None, repr=False)
    shots: Optional[dict] = field(default=None, repr=False)
    attacks: Optional[dict] = field(default=None, repr=False)

    def winner(self) -> Optional[str]:
        """Return 'home', 'away', 'draw' or None if score missing."""
        if self.home_goal is None or self.away_goal is None:
            return None
        if self.home_goal > self.away_goal:
            return "home"
        if self.away_goal > self.home_goal:
            return "away"
        return "draw"

    def to_dict(self) -> dict:
        return {
            "competition": self.competition,
            "season": self.season,
            "date": self.date.strftime("%Y-%m-%d") if self.date else None,
            "home_team": display_name(self.home),
            "away_team": display_name(self.away),
            "home_goal": self.home_goal,
            "away_goal": self.away_goal,
            "round": self.round,
            "stage": self.stage,
            "stadium": self.stadium,
            "source_file": self.source_file,
        }


def _read_csv(path: Path) -> list[dict]:
    rows: list[dict] = []
    with open(path, encoding="utf-8-sig", newline="") as fh:
        for row in csv.DictReader(fh):
            rows.append(row)
    return rows


def _load_brasileirao(path: Path) -> list[Match]:
    matches = []
    for row in _read_csv(path):
        matches.append(
            Match(
                competition=BRASILEIRAO,
                season=_to_int(row.get("season")),
                date=parse_date(row.get("datetime", "")),
                home=row.get("home_team", "").strip(),
                away=row.get("away_team", "").strip(),
                home_goal=_to_int(row.get("home_goal")),
                away_goal=_to_int(row.get("away_goal")),
                round=str(row.get("round", "") or "") or None,
                home_state=row.get("home_team_state", "").strip() or None,
                away_state=row.get("away_team_state", "").strip() or None,
                source_file=path.name,
            )
        )
    return matches


def _load_copa_do_brasil(path: Path) -> list[Match]:
    matches = []
    for row in _read_csv(path):
        matches.append(
            Match(
                competition=COPA_DO_BRASIL,
                season=_to_int(row.get("season")),
                date=parse_date(row.get("datetime", "")),
                home=row.get("home_team", "").strip(),
                away=row.get("away_team", "").strip(),
                home_goal=_to_int(row.get("home_goal")),
                away_goal=_to_int(row.get("away_goal")),
                round=row.get("round", "").strip() or None,
                source_file=path.name,
            )
        )
    return matches


def _load_libertadores(path: Path) -> list[Match]:
    matches = []
    for row in _read_csv(path):
        matches.append(
            Match(
                competition=LIBERTADORES,
                season=_to_int(row.get("season")),
                date=parse_date(row.get("datetime", "")),
                home=row.get("home_team", "").strip(),
                away=row.get("away_team", "").strip(),
                home_goal=_to_int(row.get("home_goal")),
                away_goal=_to_int(row.get("away_goal")),
                stage=row.get("stage", "").strip() or None,
                source_file=path.name,
            )
        )
    return matches


def _load_br_dataset(path: Path) -> list[Match]:
    matches = []
    for row in _read_csv(path):
        d = parse_date(row.get("date", ""))
        corners = shots = attacks = None
        hc, ac = _to_int(row.get("home_corner")), _to_int(row.get("away_corner"))
        hs, as_ = _to_int(row.get("home_shots")), _to_int(row.get("away_shots"))
        ha, aa = _to_int(row.get("home_attack")), _to_int(row.get("away_attack"))
        if hc is not None or ac is not None:
            corners = {"home": hc, "away": ac}
        if hs is not None or as_ is not None:
            shots = {"home": hs, "away": as_}
        if ha is not None or aa is not None:
            attacks = {"home": ha, "away": aa}
        matches.append(
            Match(
                competition=(row.get("tournament") or "").strip() or "Unknown",
                season=d.year if d else None,
                date=d,
                home=row.get("home", "").strip(),
                away=row.get("away", "").strip(),
                home_goal=_to_int(row.get("home_goal")),
                away_goal=_to_int(row.get("away_goal")),
                source_file=path.name,
                corners=corners,
                shots=shots,
                attacks=attacks,
            )
        )
    return matches


def _load_historico(path: Path) -> list[Match]:
    matches = []
    for row in _read_csv(path):
        matches.append(
            Match(
                competition=HISTORICO,
                season=_to_int(row.get("Ano")),
                date=parse_date(row.get("Data", "")),
                home=row.get("Equipe_mandante", "").strip(),
                away=row.get("Equipe_visitante", "").strip(),
                home_goal=_to_int(row.get("Gols_mandante")),
                away_goal=_to_int(row.get("Gols_visitante")),
                round=str(row.get("Rodada", "") or "") or None,
                home_state=row.get("Mandante_UF", "").strip() or None,
                away_state=row.get("Visitante_UF", "").strip() or None,
                stadium=row.get("Arena", "").strip() or None,
                source_file=path.name,
            )
        )
    return matches


_PLAYER_INT_COLS = {
    "ID", "Age", "Overall", "Potential", "Jersey Number",
    "International Reputation", "Weak Foot", "Skill Moves",
}
_PLAYER_FLOAT_COLS = {
    "Value", "Wage", "Height", "Weight", "Crossing", "Finishing",
    "HeadingAccuracy", "ShortPassing", "Volleys", "Dribbling", "Curve",
    "FKAccuracy", "LongPassing", "BallControl", "Acceleration",
    "SprintSpeed", "Agility", "Reactions", "Balance", "ShotPower",
    "Jumping", "Stamina", "Strength", "LongShots", "Aggression",
    "Interceptions", "Positioning", "Vision", "Penalties", "Composure",
    "Marking", "StandingTackle", "SlidingTackle", "GKDiving", "GKHandling",
    "GKKicking", "GKPositioning", "GKReflexes",
}


def _load_players(path: Path) -> list[dict]:
    players = []
    for row in _read_csv(path):
        rec: dict = {}
        for key, raw in row.items():
            if key is None:
                continue
            key = key.strip()
            if key in _PLAYER_INT_COLS:
                v = _to_int(raw)
            elif key in _PLAYER_FLOAT_COLS:
                try:
                    v = float(str(raw).strip()) if str(raw).strip() else None
                except ValueError:
                    v = None
            else:
                v = (raw or "").strip() or None
            rec[key] = v
        rec["_norm_name"] = normalize_name(rec.get("Name") or "")
        rec["_norm_club"] = normalize_name(rec.get("Club") or "")
        players.append(rec)
    return players


class SoccerData:
    """Container that loads and indexes all datasets."""

    def __init__(self, data_dir: Path | str = DEFAULT_DATA_DIR):
        self.data_dir = Path(data_dir)

        self.matches: list[Match] = []
        self.players: list[dict] = []

        self.matches.extend(self._safe("Brasileirao_Matches.csv", _load_brasileirao))
        self.matches.extend(self._safe("Brazilian_Cup_Matches.csv", _load_copa_do_brasil))
        self.matches.extend(self._safe("Libertadores_Matches.csv", _load_libertadores))
        self.matches.extend(self._safe("BR-Football-Dataset.csv", _load_br_dataset))
        self.matches.extend(self._safe("novo_campeonato_brasileiro.csv", _load_historico))
        self.players.extend(self._safe("fifa_data.csv", _load_players))

    def _safe(self, filename: str, loader) -> list:
        path = self.data_dir / filename
        if not path.exists():
            return []
        return loader(path)

    # ------------------------------------------------------------------
    # lookup helpers
    # ------------------------------------------------------------------
    def competition_matches(
        self, competition: Optional[str] = None, season: Optional[int] = None
    ) -> list[Match]:
        """Matches for a competition and season, deduplicated by source.

        Several CSV files cover the same real competition+season (e.g.
        three files contain the 2019 Brasileirão). To avoid double
        counting, only matches from the single best source file per
        season are returned. When no competition is given, all matches
        are returned as-is.
        """
        if competition is None:
            out = list(self.matches)
        else:
            group = competition_group(competition)
            out = [m for m in self.matches if normalize_name(m.competition) in group]
        if season is not None:
            out = [m for m in out if m.season == season]
            if competition is not None:
                return self._best_source_for(out, season)
            return out
        if competition is None:
            return out
        seasons = sorted({m.season for m in out if m.season is not None})
        result: list[Match] = []
        for s in seasons:
            result.extend(self._best_source_for(out, s))
        result.extend(m for m in out if m.season is None)
        return result

    def _best_source_for(self, matches: list[Match], season: int) -> list[Match]:
        by_file: dict[str, list[Match]] = defaultdict(list)
        for m in matches:
            if m.season == season:
                by_file[m.source_file or ""].append(m)
        if not by_file:
            return []
        best = min(
            by_file.items(),
            key=lambda kv: (-len(kv[1]), self._source_rank(kv[0])),
        )
        return best[1]

    @staticmethod
    def _source_rank(source_file: str) -> int:
        try:
            return _SOURCE_PRIORITY.index(source_file)
        except ValueError:
            return len(_SOURCE_PRIORITY)

    def team_matches(self, team: str) -> list[Match]:
        """All matches involving a team (home or away), any competition."""
        return [
            m
            for m in self.matches
            if name_matches(team, m.home) or name_matches(team, m.away)
        ]

    def head_to_head(self, team_a: str, team_b: str) -> list[Match]:
        """All matches where the two teams faced each other."""
        out = []
        for m in self.matches:
            a_home = name_matches(team_a, m.home)
            a_away = name_matches(team_a, m.away)
            b_home = name_matches(team_b, m.home)
            b_away = name_matches(team_b, m.away)
            if (a_home or b_home) and (a_away or b_away):
                # both teams present, on opposite sides
                if (a_home or a_away) and (b_home or b_away):
                    out.append(m)
        out.sort(key=lambda m: (m.date or datetime.min))
        return out

    def find_player(self, name: str) -> list[dict]:
        n = normalize_name(name)
        return [p for p in self.players if n in (p.get("_norm_name") or "")]

    def club_players(self, club: str) -> list[dict]:
        c = normalize_name(club)
        return [p for p in self.players if p.get("_norm_club") == c]

    def competitions(self) -> list[str]:
        seen = []
        for m in self.matches:
            if m.competition not in seen:
                seen.append(m.competition)
        return seen
